src/rust/crate/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+//!
+//! This crate allows to create a context with which it's possible to sign and verify messages to communicate with other Sessionless instances.
+//!
+//! Enable the `uuid` feature to be able to generate random UUIDs _(useful for the server)_.
+
+pub extern crate secp256k1;
+
+#[cfg(test)]
+mod tests;
+
+/// Module for the hex encoding.
+pub mod hex;
+
+use secp256k1::{Secp256k1, All, Message, rand::rngs::OsRng};
+use sha3::{Digest, Keccak256, digest::{Output}, Keccak256Core};
+
+pub type Signature = secp256k1::ecdsa::Signature;
+pub type PrivateKey = secp256k1::SecretKey;
+pub type PublicKey = secp256k1::PublicKey;
+
+/// Sessionless context structure used to perform all signature operations.
+/// It contains the `secp256k1` engine and private-public key pair.
+///
+/// Do not create a new context object for each operation, as construction is
+/// far slower than all other API calls.
+pub struct Sessionless {
+    ctx: Secp256k1<All>,
+    private_key: PrivateKey,
+    public_key: PublicKey,
+}
+
+impl Sessionless {
+    /// Creates a new randomized Sessionless context.
+    ///
+    /// Private key is generated using `OsRng`.
+    pub fn new() -> Self {
+        let ctx = Secp256k1::new();
+        let (private_key, public_key) = ctx.generate_keypair(&mut OsRng);
+
+        Self {
+            ctx,
+            private_key,
+            public_key,
+        }
+    }
+
+    /// Creates a randomized Sessionless context, but with a designated key pair.
+    pub fn from_private_key(private_key: PrivateKey) -> Self {
+        let ctx = Secp256k1::new();
+        let public_key = private_key.public_key(&ctx);
+
+        Self {
+            ctx,
+            private_key,
+            public_key,
+        }
+    }
+
+    #[cfg(test)]
+    /// Creates a Sessionless context with predefined output.
+    ///
+    /// Used ONLY for testing purposes.
+    pub fn new_predefined() -> Self {
+        let mut ctx = Secp256k1::new();
+        ctx.seeded_randomize(&[0u8; 32]);
+
+        let mut step_rng = secp256k1::rand::rngs::mock::StepRng::new(0, 1);
+        let (private_key, public_key) = ctx.generate_keypair(&mut step_rng);
+
+        Self {
+            ctx,
+            private_key,
+            public_key,
+        }
+    }
+
+    /// Constructs a signature for `message` using the context's secret key.
+    pub fn sign(&self, message: impl AsRef<[u8]>) -> Signature {
+        let message_hash = self.hash_message(message);
+        self.ctx.sign_ecdsa(&message_hash, &self.private_key).into()
+    }
+
+    /// Checks if the `signature` is valid for the `message` using the `public key`
+    /// which was a part of the context used to construct that signature.
+    pub fn verify(
+        &self,
+        message: impl AsRef<[u8]>,
+        public_key: &PublicKey,
+        signature: &Signature,
+    ) -> Result<(), secp256k1::Error> {
+        let message_hash = self.hash_message(message);
```

Comment:
The `hash_message` function is defined bellow and works exactly the same as in the JS implementation as it is the same algorithm (keccak256) and no slicing is done in the [`sign` JS function](https://github.com/planet-nine-app/sessionless/blob/d630759e16f8049faa45ccf90a8ca3159eb9ce68/src/javascript/node/sessionless.js#L34). I even made a test. I've hashed message `"1234567890_1234567890_1234567890_1234567890"` in both implementations, here are the results:

Java Script:
```js
import { keccak256 } from "ethereum-cryptography/keccak.js";
import { utf8ToBytes } from "ethereum-cryptography/utils.js";

const messageHash = keccak256(utf8ToBytes("1234567890_1234567890_1234567890_1234567890"));

console.log(messageHash);
```
Output:
```
Uint8Array(32) [
   25,   7, 137,  94, 228, 243, 187,  89,
  147, 149,  87, 149, 153,  72,  78, 141,
  127, 216, 118, 254, 112, 170, 187, 108,
  201, 220,   2,  49, 107, 225, 148,  97
]
```

Rust:
```rs
let mut hasher = Keccak256::new();
hasher.update(b"1234567890_1234567890_1234567890_1234567890");
let digest: Output<Keccak256Core> = hasher.finalize();

println!("{:?}", &*digest)
```
Output:
```
[25, 7, 137, 94, 228, 243, 187, 89, 147, 149, 87, 149, 153, 72, 78, 141, 127, 216, 118, 254, 112, 170, 187, 108, 201, 220, 2, 49, 107, 225, 148, 97]
```

As you can see both produce the same hash.

I also don't think that slicing the message is a good idea. Let's say we have 2 different messages:
- `1234567890_1234567890_1234567890_1234567890`
- `1234567890_1234567890_1234567890_0000000000`

If we now slice them to 32 bytes we would have:
- `1234567890_1234567890_1234567890`
- `1234567890_1234567890_1234567890`

Which is the same input that would produce the same hash for 2 different messages.

If you agree and additionally want to standardize the Signature format to hex string instead of JSON structure I think we can merge and then when it's time to do something with the UUIDs and Public Keys we can just update the implementation.

Reply:
yep totally agree. I was misremembering. This: `Uint8Array(32)` produces a 32 byte _hash_ not a 32 char message. So yeah let's merge this and standardize as part of the demo we put together.